Add generators for SMBIOS type 4 and 7 tables (CPU and cache topologies)

sarah-walker-arm opened the pull request.

sarah-walker-arm committed:
DynamicTablesPkg: Smbios Processor Information (Type 4)

Signed-off-by: Sarah Walker <Sarah.Walker2@arm.com>

sarah-walker-arm committed:
DynamicTablesPkg: Implement abstract CM_OBJECT_TOKENs

Abstract tokens simplify linking of generated tables. They allow for tables
to be search for not just by existing ConfigurationManager object references,
but also by arbitary tokens generated by the table generators themselves.

Signed-off-by: Sarah Walker <Sarah.Walker2@arm.com>

sarah-walker-arm committed:
MdePkg: Smbios: Add AArch64 ProcessorId variant for type 4 table

The existing field is x86 only. Re-typedef for AArch64.

Signed-off-by: Sarah Walker <Sarah.Walker2@arm.com>

sarah-walker-arm committed:
MdePkg: SmBios: Add structs for cache size and configuration data

These are used by the Type 7 generator.

Signed-off-by: Sarah Walker <Sarah.Walker2@arm.com>

sarah-walker-arm committed:
DynamicTablesPkg: Smbios Cache Information (Type 7)

Signed-off-by: Sarah Walker <Sarah.Walker2@arm.com>

mgamal reviewed (approved):
Looks good to me.

pierregondois commented on DynamicTablesPkg/Include/StandardNameSpaceObjects.h:
```diff
@@ -40,6 +48,45 @@
+Object ID's in the SMBIOS Namespace:
+   See ESTD_SMBIOS_TABLE_ID.
+*/
+#define CM_ABSTRACT_TOKEN_MAKE(ns, obj_id, id)  ((CM_OBJECT_TOKEN)(((ns) << 28) | ((obj_id) << 16) | ((id) << 1) | 1))
```
Shouldn't the namespace bits be shifter by 24 ?
Also NIT, but maybe we could mask the bits after the index 15 to be sure not to override the object id bits

pierregondois commented on DynamicTablesPkg/Include/StandardNameSpaceObjects.h:
```diff
@@ -27,9 +27,17 @@
+  previously generated tables. To aid with this, the concept
+  of an abstract token exists, allowing generators to
+  define arbitary tokens to identify generated tables.
+  When bit 1 is set, the token is an abstract token and
```
Is it necessary to set the first bit ?
Most tokens are currently addresses, so setting the first bit should not have a big impact. I assume having the higher bits set to `0b1010` is more important.

sarah-walker-arm replied:
As it would seem likely that all token addresses would at least be word aligned, setting bit 0 seems a reasonable way to guarantee that there are no clashes between normal and abstract tokens. Higher bits won't necessarily achieve that.

pierregondois replied:
Ah ok yes, thanks for the explanation

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,691 @@
+      StringTablePublishStringSet (&StrTable, (CHAR8 *)(SmbiosRecord + 1), SmbiosRecordSize - sizeof (SMBIOS_TABLE_TYPE7));
+
+      TableList[ObjIndex]          = (SMBIOS_STRUCTURE *)SmbiosRecord;
+      CmObjectList[ObjIndex]       = CM_ABSTRACT_TOKEN_MAKE (ETokenNameSpaceSmbios, EStdSmbiosTableIdType07, CacheNode->Level | (SocketIndex << 2));
```
Just to share some thoughts, ideally we should have a common way to create abstract token values. Meaning that the logic to create the abstract token `CacheNode->Level | (SocketIndex << 2` works because there are not so many objects to create.
Ideally I think we should rely on the MetadataLib and a MetadataObject like the following:
```
typedef struct MetadataObjProcCache {
  /// Token of a CM_ARCH_COMMON_PROC_HIERARCHY_INFO
  UINT32    Token;

  UINT32    Level1CacheSmbiosHandle;
  UINT32    Level2CacheSmbiosHandle;
  UINT32    Level3CacheSmbiosHandle;
} METADATA_OBJ_PROC_CACHE;
```

instead of abstract handles and this set of callbacks:
-  AddSmbiosHandle()
-  FindSmbiosHandle()
-  FindSmbiosHandleEx()
but this would require more modifications.

I assume this can be discussed later as what I suggest seems un-necessary right now.
I ll finish the code review this week, sorry for the delay

sarah-walker-arm replied:
As the documentation in StandardNameSpaceObjects.h states, the ID values are arbitrary and are defined separately for each name space / object ID pair. As a result there is no common way to create abstract token values, as it depends on the object type.

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Lib.inf:
```diff
@@ -0,0 +1,36 @@
+  DynamicTablesPkg/DynamicTablesPkg.dec
+
+[Protocols]
+  gEfiSmbiosProtocolGuid                        # PROTOCOL ALWAYS_CONSUMED
```
Could remove I think

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Lib.inf:
```diff
@@ -0,0 +1,36 @@
+[Defines]
+  INF_VERSION    = 0x0001001B
+  BASE_NAME      = SmbiosType7LibArm
+  FILE_GUID      = a256956a-de0b-4aa7-8eac-5ce13bbfbb1f
```
Same GUID as in table4 generator

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,691 @@
+#include <ConfigurationManagerHelper.h>
+#include <Protocol/ConfigurationManagerProtocol.h>
+#include <Protocol/DynamicTableFactoryProtocol.h>
+#include <Protocol/Smbios.h>
```
Could remove I think

pierregondois commented on DynamicTablesPkg/Include/ArchCommonNameSpaceObjects.h:
```diff
@@ -451,6 +451,10 @@ typedef struct CmArchCommonCacheInfo {
   UINT16             LineSize;
   /// Unique ID for the cache
   UINT32             CacheId;
+  /// Level of cache within the processor hierarchy
```
It would be good to tag the fields that are specific to SMBIOS tables only. Maybe: 
```
/// SMBIOS: Level of ...
```
or something similar

pierregondois commented on DynamicTablesPkg/Include/ArchCommonNameSpaceObjects.h:
```diff
@@ -451,6 +451,10 @@ typedef struct CmArchCommonCacheInfo {
+  /// Level of cache within the processor hierarchy
+  UINT32             Level;
+  /// Designation of this cache on this socket
+  CHAR8              *SocketDesignation;
```
I think this means we won't be able to add new fields after this one as this has a variable size. Would it be possible to define a maximum size for the strings instead ? Something like:
```
#define MAXIMUM_SMBIOS_STRING_SIZE   ???
[...]
SocketDesignation[MAXIMUM_SMBIOS_STRING_SIZE];
```

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,691 @@
+STATIC
+CACHE_TYPE_DATA
+GetSystemCacheType (
+  IN CONST UINT32  Attributes
```
NIT: no need for CONST I think, same comment for `GetCacheAssociativity()`

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,691 @@
+  for (Index = 0; Index < ProcHierarchyNodeCount; Index++) {
+    Node = &ProcHierarchyNodeList[Index];
+
+    Status = CfgMgrProtocol->GetObject (CfgMgrProtocol, CREATE_CM_ARCH_COMMON_OBJECT_ID (EArchCommonObjCmRef), Node->PrivateResourcesArrayToken, &CmObjectDesc);
```
NIT: to avoid too many levels of indentation maybe (and to catch errors other than `EFI_NOT_FOUND`):

```
Status = CfgMgrProtocol->GetObject ()
if (Status == EFI_NOT_FOUND) {
  continue;
} else if (EFI_ERROR (Status)) {
  ASSERT_EFI_ERROR (Status);
  // return -1 ?
}
...
```

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,691 @@
+    ProcNode = &ProcHierarchyNodeList[Index];
+    Status   = CfgMgrProtocol->GetObject (CfgMgrProtocol, CREATE_CM_ARCH_COMMON_OBJECT_ID (EArchCommonObjCmRef), ProcNode->PrivateResourcesArrayToken, &CmObjectDesc);
+
+    if (Status == EFI_SUCCESS) {
```
Similar comment about trying to avoid too many levels of indentation

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,691 @@
+      PrivResources = (CM_ARCH_COMMON_OBJ_REF *)CmObjectDesc.Data;
+
+      for (ObjIndex = 0; ObjIndex < CmObjectDesc.Count; ObjIndex++) {
+        if (PrivResources[ObjIndex].ReferenceToken == CacheNode->Token) {
```
Could also revert the condition here:
```
if (PrivResources[ObjIndex].ReferenceToken != CacheNode->Token) {
  continue;
}
```

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,691 @@
+          // Walk the processor hierarchy to find the socket
+          while (!(ProcNode->Flags & EFI_ACPI_6_3_PPTT_PACKAGE_PHYSICAL)) {
+            ProcNode = FindProcHierarchyInfoFromToken (ProcHierarchyNodeList, ProcHierarchyNodeCount, ProcNode->ParentToken);
+            ASSERT (ProcNode != NULL);
```
Maybe we should also return NULL here, in case this is a RELEASE build

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,691 @@
+  UINTN                               Index;
+  UINTN                               ObjIndex;
+
+  // Search for a processor hierarchy node using this cache
```
I think there might be an issue here, as a cache node might be referenced by processors belonging to 2 different sockets. If we have 2 sockets using the same cache nodes, we won't associate any cache to the second socket IIUC.
I think we might need to first find the number of sockets on the platform, and then call this function to check if a cache node can be associated to each socket.

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,691 @@
+      __FUNCTION__,
+      Status
+      ));
+    return EFI_INVALID_PARAMETER;
```
Shouldn't we return the initial Status ?

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,691 @@
+             &CacheStructList,
+             &CacheStructCount
+             );
+  if (EFI_ERROR (Status) && (Status != EFI_NOT_FOUND)) {
```
If not found or if `CacheStructCount == 0`, I assume we could just return here (or signal there was nothing to do)

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,691 @@
+  *TableCount    = ObjIndex;
+  FreePool (SocketCmObjectList);
+  Status = EFI_SUCCESS;
+exitBuildSmbiosType7Table:
```
Should we just return instead of going to this label, as there is no need to free any memory ?

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,691 @@
+
+      SmbiosRecord->MaximumCacheSize.Granularity64K = 1;
+      SmbiosRecord->MaximumCacheSize.Size           = 0x7fff;
+    } else if (CacheSize > 32767) {
```
I m not sure I understand where this value is coming from

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,691 @@
+      SmbiosRecord     = (SMBIOS_TABLE_TYPE7 *)AllocateZeroPool (SmbiosRecordSize);
+
+      // Set up the header
+      SmbiosRecord->Hdr.Type   = EFI_SMBIOS_TYPE_CACHE_INFORMATION;
```
```
6.1.3 Text strings
NOTE There is no limit on the length of each individual text string. However, the length of the entire structure table
(including all strings) must be reported in the Structure Table Length field of the 32-bit Structure Table Entry Point
(see 5.2.1) and/or the Structure Table Maximum Size field of the 64-bit Structure Table Entry Point (see 5.2.2).
```
I think the length of the table should be increased

sarah-walker-arm replied:
I believe that is referring to the length of the SMBIOS tables _as a whole_. My reading is that the length field for each record refers only to the fixed length portion, and does not include the strings; the length field is a uint8_t which wouldn't allow for many strings anyway.

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,691 @@
+      if (CacheNode->SocketDesignation) {
+        Status = StringTableAddString (&StrTable, CacheNode->SocketDesignation, &SocketDesignationRef);
+        if (EFI_ERROR (Status)) {
+          DEBUG ((DEBUG_ERROR, "Failed to add Socket Designation String %r\n", Status));
```
Should we assert/return if there is an error ?

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,691 @@
+
+      SmbiosRecord->MaximumCacheSize2.Size += CacheSize / 1024;
+
+      if (GetSystemCacheType (CacheNode->Attributes) != SmbiosRecord->SystemCacheType) {
```
Shouldn't this be more complex ? I.e.:
- Data + Instr. -> Unified
- Instr. + Data -> Unified
- Other combinations: Others, (or even assert just to see if someone is actually using this kind of topology)

pierregondois commented on DynamicTablesPkg/Include/ArchCommonNameSpaceObjects.h:
```diff
@@ -421,6 +421,24 @@ typedef struct CmArchCommonProcHierarchyInfo {
+  /// Designation of this CM_ARCH_COMMON_PROC_HIERARCHY_INFO instance.
+  /// This string (and all that follow) are intended only for instances with
+  /// EFI_ACPI_6_3_PPTT_PACKAGE_PHYSICAL set, ie describing physical sockets.
+  CHAR8              *SocketDesignation;
```
Same comments as for the Type7 table (just a note in case):
- Is it possible to tag these new fields as `/// SMBIOS: ...`
- Is it possible to have a fixed-size array instead of a pointer ?

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType4Lib/SmbiosType4Lib.inf:
```diff
@@ -0,0 +1,36 @@
+  ArmPlatformPkg/ArmPlatformPkg.dec
+  DynamicTablesPkg/DynamicTablesPkg.dec
+
+[Protocols]
```
Not necessary I think

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType4Lib/SmbiosType4Generator.c:
```diff
@@ -0,0 +1,538 @@
+#include <ConfigurationManagerHelper.h>
+#include <Protocol/ConfigurationManagerProtocol.h>
+#include <Protocol/DynamicTableFactoryProtocol.h>
+#include <Protocol/Smbios.h>
```
Not necessary I think

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType4Lib/SmbiosType4Generator.c:
```diff
@@ -0,0 +1,538 @@
+    }
+  }
+
+  ASSERT (SocketCount != 0);
```
Maybe we should also return here in case this is a RELEASE build

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType4Lib/SmbiosType4Generator.c:
```diff
@@ -0,0 +1,538 @@
+  ObjIndex = 0;
+
+  for (Index = 0; Index < ProcHierarchyNodeCount; Index++) {
+    if (ProcHierarchyNodeList[Index].Flags & EFI_ACPI_6_3_PPTT_PACKAGE_PHYSICAL) {
```
Could revert the condition to be one level of indentation lower:
```
if ((ProcHierarchyNodeList[Index].Flags & EFI_ACPI_6_3_PPTT_PACKAGE_PHYSICAL) == 0) {
  continue;
}
```

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType4Lib/SmbiosType4Generator.c:
```diff
@@ -0,0 +1,538 @@
+      if (ProcHierarchyNodeList[Index].SocketDesignation) {
+        Status = StringTableAddString (&StrTable, ProcHierarchyNodeList[Index].SocketDesignation, &SocketDesignationRef);
+        if (EFI_ERROR (Status)) {
+          DEBUG ((DEBUG_ERROR, "Failed to add Socket Designation String %r\n", Status));
```
Should we assert in case of error ? I didn't see that these string fields were optional, so I assume this should be an error no ?

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType4Lib/SmbiosType4Generator.c:
```diff
@@ -0,0 +1,538 @@
+
+      // Set up the header
+      SmbiosRecord->Hdr.Type   = EFI_SMBIOS_TYPE_PROCESSOR_INFORMATION;
+      SmbiosRecord->Hdr.Length = sizeof (SMBIOS_TABLE_TYPE4);
```
Similar comment as for the Type7: shouldn't this be set to SmbiosRecordSize ?

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType4Lib/SmbiosType4Generator.c:
```diff
@@ -0,0 +1,538 @@
+  *CmObjectToken = NULL;
+  *TableCount    = SocketCount;
+  Status         = EFI_SUCCESS;
+exitBuildSmbiosType4Table:
```
Do we need the label as there is nothing to free ?

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType4Lib/SmbiosType4Generator.c:
```diff
@@ -0,0 +1,538 @@
+          // Walk through ParentToken to find the socket node
+          while (!Flags->PhysicalPackage) {
+            Node = FindProcHierarchyInfoFromToken (ProcHierarchyNodeList, ProcHierarchyNodeCount, Node->ParentToken);
+            ASSERT (Node != NULL);
```
Maybe we should also have a return statement here for RELEASE builds

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType4Lib/SmbiosType4Generator.c:
```diff
@@ -0,0 +1,538 @@
+          }
+
+          // Is this CPU on the current socket?
+          if (Node == &ProcHierarchyNodeList[Index]) {
```
Could revert some conditions to lower the indentation

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType4Lib/SmbiosType4Generator.c:
```diff
@@ -0,0 +1,538 @@
+ #if defined (MDE_CPU_AARCH64)
+      SmbiosRecord->ProcessorFamily          = ProcessorFamilyIndicatorFamily2;
+      SmbiosRecord->ProcessorFamily2         = ProcessorFamilyARMv8;
+      SmbiosRecord->ProcessorId.SocId        = ProcHierarchyNodeList[Index].ProcessorId & 0xffff;
```
We could replace the `ProcessorId` type by PROCESSOR_ID_DATA and copy paste the field at this point

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType4Lib/SmbiosType4Generator.c:
```diff
@@ -0,0 +1,538 @@
+ #else
+      #error SmbiosType4Generator needs extending for your CPU
+ #endif
+
```
NIT: Maybe we should also set:
```
ProcessorUpgrade = ProcessorUpgradeUnknown
```

pierregondois commented:
Hi Sarah,
Thanks for all the update, I think there should be only one comment left

pierregondois commented on MdePkg/Include/IndustryStandard/SmBios.h:
```diff
@@ -975,10 +975,19 @@ typedef union {
+  UINT8     SipId;
+  UINT8     SipBankIndex;
+  UINT32    SocRevision;
+} PROCESSOR_ID_DATA;
```
I think this will also require modifications in ArmPkg/edk2-platforms, some tables have been hardcoded and don't seem to compile anymore with this field modification.
Same comment for the `CACHE_SIZE` and `CACHE_SIZE_2` modifications, I think we will have to update the field assignment at every place `SMBIOS_TABLE_TYPE4` and `SMBIOS_TABLE_TYPE7` struct are used

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,804 @@
+  for (Index = 0; Index < ProcHierarchyNodeCount; Index++) {
+    Node = &ProcHierarchyNodeList[Index];
+
+    Status = CfgMgrProtocol->GetObject (CfgMgrProtocol, CREATE_CM_ARCH_COMMON_OBJECT_ID (EArchCommonObjCmRef), Node->PrivateResourcesArrayToken, &CmObjectDesc);
```
Is it possible to use the following as the macro contains some additional checks ?
```
GET_OBJECT_LIST (
  EObjNameSpaceArchCommon,
  EArchCommonObjCmRef,
  CM_ARCH_COMMON_OBJ_REF
  );
...

Status = GetEArchCommonObjCmRef (
                CfgMgrProtocol,
                Node->PrivateResourcesArrayToken,
                &PrivResources,
                &PrivResourcesCount
                );
```

Same comment for other places where `CfgMgrProtocol->GetObject()` is used.

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,804 @@
+        StringTableInitialize (&StrTable, SMBIOS_TYPE7_MAX_STRINGS);
+        SocketDesignationRef = 0;
+
+        if (CacheNode->SocketDesignation) {
```
Now that Strings are arrays instead of pointers, I think this should become:
`if (CacheNode->SocketDesignation[0] != 0) {`
(same comments for other strings)

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType4Lib/SmbiosType4Generator.c:
```diff
@@ -0,0 +1,573 @@
+  ASSERT (ObjIndex == SocketCount);
+
+  *Table         = TableList;
+  *CmObjectToken = NULL;
```
In `DynamicTablesPkg/Drivers/DynamicTableManagerDxe/SmbiosTableBuilder.c`:
```
Status = TableFactoryProtocol->AddSmbiosHandle (
                                 SmbiosProtocol,
                                 &TableHandle,
                                 CmObjToken[Index],
                                 SmbiosTableInfo->TableGeneratorId
                                 );
```

we access `CmObjToken[Index]` withouth checking for a NULL array, so I think an additional patch is required for that

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType4Lib/SmbiosType4Generator.c:
```diff
@@ -0,0 +1,573 @@
+    SmbiosRecord->ProcessorVersion      = ProcessorVersionRef;
+    SmbiosRecord->SerialNumber          = SerialNumberRef;
+    SmbiosRecord->AssetTag              = AssetTagRef;
+    SmbiosRecord->PartNumber            = PartNumberRef;
```
Similar comment as for the L1-Data/L1-Instruction cache split.

The Juno has big/little CPUs in the same socket. In the current Smbios tables of the platform, they have distinct Type4 tables and Type7 tables (little cluster L2 and big cluster L2 are distinct tables).

What you implemented seems correct spec. wise, but people chose to do it differently and IMO it also makes sense to split the type4 table here, so I m not sure what to say here...

sarah-walker-arm replied:
Unfortunately the type 4 spec is quite clear that each structure instance corresponds to a socket/slot. So the existing smbios tables are wrong.

pierregondois commented on DynamicTablesPkg/Include/ArchCommonNameSpaceObjects.h:
```diff
@@ -451,6 +453,10 @@ typedef struct CmArchCommonCacheInfo {
   /// Unique ID for the cache
   UINT32             CacheId;
+  /// SMBIOS: Level of cache within the processor hierarchy
+  UINT32             Level;
```
not sure if it is better to do `-1` in the code or to add a not here saying that cache levels are computed as:
Level 1 -> 0
Level 2 -> 1
Level 3 -> 2

pierregondois commented on DynamicTablesPkg/Library/Smbios/SmbiosType7Lib/SmbiosType7Generator.c:
```diff
@@ -0,0 +1,804 @@
+
+        SmbiosRecord->MaximumCacheSize2.Size += CacheSize / 1024;
+
+        if (((GetSystemCacheType (CacheNode->Attributes) == CacheTypeData) &&
```
When reading the spec. this seems correct.
For the Juno and other platforms, it seems people implemented it differently, i.e. there is one table for the `L1 data cache` and one table for the `L1 instruction cache`.
So I m not sure what should be done here ...

sarah-walker-arm committed:
DynamicTablesPkg: DynamicTableManagerDxe: Fix NULL pointer dereference

BuildAndInstallMultipleSmbiosTables() can dereference a NULL pointer if the
SMBIOS table builder returns a NULL CmObjToken array.

Signed-off-by: Sarah Walker <Sarah.Walker2@arm.com>

sarah-walker-arm committed:
DynamicTablesPkg: Add SmbiosSmcLib

Implement a support library for SMBIOS-related SMC calls. Currently this
implements a function to return the SoC ID.

Signed-off-by: Sarah Walker <Sarah.Walker2@arm.com>

pierregondois commented:
Some final comments:
- `File Include/Library/SmbiosSmcLib.h not declared in package DEC DynamicTablesPkg/DynamicTablesPkg.dec
`
- I got an uncrustify error
- I think a patch for the ArmPkg should be added, cf. the `PROCESSOR_ID_DATA` struct being updated

sarah-walker-arm committed:
ArmPkg: Smbios: Update ProcessorSubClassDxe for new SMBIOS structures

Update ProcessorSubClassDxe to handle the new cache size and AArch64
PROCESSOR_ID_DATA SMBIOS structures.

Signed-off-by: Sarah Walker <Sarah.Walker2@arm.com>
